File: whatsapp-bridge/agent_logger.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"sync"
	"time"
)

// AgentLogEntry represents a single log entry in the agent's decision-making process
type AgentLogEntry struct {
	Timestamp      time.Time              `json:"timestamp"`
	ChatJID        string                 `json:"chat_jid"`
	ChatName       string                 `json:"chat_name,omitempty"`
	Stage          string                 `json:"stage"`           // "received", "analysis", "decision", "response"
	MessageID      string                 `json:"message_id,omitempty"`
	Sender         string                 `json:"sender,omitempty"`
	MessageContent string                 `json:"message_content,omitempty"`
	Logic          map[string]interface{} `json:"logic,omitempty"` // Agent's reasoning and analysis
	Response       string                 `json:"response,omitempty"`
	Error          string                 `json:"error,omitempty"`
}

// AgentLogger handles structured logging of agent interactions and decision-making
type AgentLogger struct {
	logDir string
	mu     sync.Mutex
}

// NewAgentLogger creates a new agent logger
func NewAgentLogger(baseDir string) (*AgentLogger, error) {
	logDir := filepath.Join(baseDir, "agent-logs")

	// Create logs directory if it doesn't exist
	if err := os.MkdirAll(logDir, 0755); err != nil {
		return nil, fmt.Errorf("failed to create agent logs directory: %w", err)
	}

	return &AgentLogger{
		logDir: logDir,
	}, nil
}

// LogReceivedMessage logs when a message is received
func (al *AgentLogger) LogReceivedMessage(chatJID, chatName, messageID, sender, content string) error {
	entry := AgentLogEntry{
		Timestamp:      time.Now(),
		ChatJID:        chatJID,
		ChatName:       chatName,
		Stage:          "received",
		MessageID:      messageID,
		Sender:         sender,
		MessageContent: content,
		Logic: map[string]interface{}{
			"message_length": len(content),
			"has_content":    content != "",
		},
	}

	return al.writeLogEntry(chatJID, entry)
}

// LogAnalysis logs the agent's analysis of the message
func (al *AgentLogger) LogAnalysis(chatJID, chatName string, analysis map[string]interface{}) error {
	entry := AgentLogEntry{
		Timestamp: time.Now(),
		ChatJID:   chatJID,
		ChatName:  chatName,
		Stage:     "analysis",
		Logic:     analysis,
	}

	return al.writeLogEntry(chatJID, entry)
}

// LogDecision logs the agent's decision (should respond, why/why not)
func (al *AgentLogger) LogDecision(chatJID, chatName string, decision map[string]interface{}) error {
	entry := AgentLogEntry{
		Timestamp: time.Now(),
		ChatJID:   chatJID,
		ChatName:  chatName,
		Stage:     "decision",
		Logic:     decision,
	}

	return al.writeLogEntry(chatJID, entry)
}

// LogResponse logs the generated response
func (al *AgentLogger) LogResponse(chatJID, chatName, messageID, response string, logic map[string]interface{}) error {
	entry := AgentLogEntry{
		Timestamp:  time.Now(),
		ChatJID:    chatJID,
		ChatName:   chatName,
		Stage:      "response",
		MessageID:  messageID,
		Response:   response,
		Logic:      logic,
	}

	return al.writeLogEntry(chatJID, entry)
}

// LogError logs an error that occurred during agent processing
func (al *AgentLogger) LogError(chatJID, chatName, stage, errorMsg string) error {
	entry := AgentLogEntry{
		Timestamp: time.Now(),
		ChatJID:   chatJID,
		ChatName:  chatName,
		Stage:     stage,
		Error:     errorMsg,
	}

	return al.writeLogEntry(chatJID, entry)
}

// LogLeaveRequest logs specifically for leave request processing with detailed logic
func (al *AgentLogger) LogLeaveRequest(chatJID, chatName, messageContent string, extractedInfo map[string]interface{}, missingFields []string, nextAction string) error {
	logic := map[string]interface{}{
		"intent":          "leave_request",
		"message_content": messageContent,
		"extracted_info":  extractedInfo,
		"missing_fields":  missingFields,
		"next_action":     nextAction,
		"analysis_detail": "",
	}

	// Build detailed analysis text
	var analysisDetail string
	if len(extractedInfo) > 0 {
		analysisDetail += "Found information:\n"
		for key, value := range extractedInfo {
			analysisDetail += fmt.Sprintf("  - %s: %v\n", key, value)
		}
	}

	if len(missingFields) > 0 {
		analysisDetail += "\nMissing required information:\n"
		for _, field := range missingFields {
			analysisDetail += fmt.Sprintf("  - %s\n", field)
		}
	}

	analysisDetail += fmt.Sprintf("\nNext action: %s\n", nextAction)
	logic["analysis_detail"] = analysisDetail

	entry := AgentLogEntry{
		Timestamp:      time.Now(),
		ChatJID:        chatJID,
		ChatName:       chatName,
		Stage:          "leave_request_analysis",
		MessageContent: messageContent,
		Logic:          logic,
	}

	return al.writeLogEntry(chatJID, entry)
}

// writeLogEntry writes a log entry to the appropriate file
func (al *AgentLogger) writeLogEntry(chatJID string, entry AgentLogEntry) error {
	al.mu.Lock()
	defer al.mu.Unlock()

	// Create a daily log file for each chat
	date := time.Now().Format("2006-01-02")
	// Sanitize JID for filename (replace @ and other special chars)
	sanitizedJID := filepath.Base(chatJID)
	logFile := filepath.Join(al.logDir, fmt.Sprintf("%s_%s.jsonl", sanitizedJID, date))

	// Open file in append mode
	f, err := os.OpenFile(logFile, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0644)
	if err != nil {
		return fmt.Errorf("failed to open log file: %w", err)
	}
	defer f.Close()

	// Write JSON line
	jsonData, err := json.Marshal(entry)
	if err != nil {
		return fmt.Errorf("failed to marshal log entry: %w", err)
	}

	if _, err := f.Write(append(jsonData, '\n')); err != nil {
		return fmt.Errorf("failed to write log entry: %w", err)
	}

	return nil
}

// GetChatLogs retrieves all log entries for a specific chat and date
func (al *AgentLogger) GetChatLogs(chatJID, date string) ([]AgentLogEntry, error) {
	sanitizedJID := filepath.Base(chatJID)
	logFile := filepath.Join(al.logDir, fmt.Sprintf("%s_%s.jsonl", sanitizedJID, date))

	data, err := os.ReadFile(logFile)
	if err != nil {
		if os.IsNotExist(err) {
			return []AgentLogEntry{}, nil // No logs for this date
		}
		return nil, fmt.Errorf("failed to read log file: %w", err)
	}

	// Parse JSONL (JSON Lines format)
	var entries []AgentLogEntry
	lines := make([]byte, 0, len(data))
	for _, b := range data {
		if b == '\n' {
			if len(lines) > 0 {
				var entry AgentLogEntry
				if err := json.Unmarshal(lines, &entry); err == nil {
					entries = append(entries, entry)
				}
				lines = lines[:0]
			}
		} else {
			lines = append(lines, b)
		}
	}

	// Handle last line if no trailing newline
	if len(lines) > 0 {
		var entry AgentLogEntry
		if err := json.Unmarshal(lines, &entry); err == nil {
			entries = append(entries, entry)
		}
	}

	return entries, nil
}

// GenerateHumanReadableLog creates a human-readable version of the logs
func (al *AgentLogger) GenerateHumanReadableLog(chatJID, date string) (string, error) {
	entries, err := al.GetChatLogs(chatJID, date)
	if err != nil {
		return "", err
	}

	if len(entries) == 0 {
		return fmt.Sprintf("No agent logs found for %s on %s\n", chatJID, date), nil
	}

	var output string
	output += fmt.Sprintf("=== Agent Logs for %s (%s) ===\n\n", chatJID, date)

	for i, entry := range entries {
		output += fmt.Sprintf("[%d] %s - Stage: %s\n", i+1, entry.Timestamp.Format("15:04:05"), entry.Stage)

		if entry.Sender != "" {
			output += fmt.Sprintf("    Sender: %s\n", entry.Sender)
		}

		if entry.MessageContent != "" {
			output += fmt.Sprintf("    Message: %s\n", entry.MessageContent)
		}

		if len(entry.Logic) > 0 {
			output += "    Logic/Analysis:\n"
			for key, value := range entry.Logic {
				if key == "analysis_detail" {
					output += fmt.Sprintf("      %v\n", value)
				} else {
					output += fmt.Sprintf("      %s: %v\n", key, value)
				}
			}
		}

		if entry.Response != "" {
			output += fmt.Sprintf("    Response: %s\n", entry.Response)
		}

		if entry.Error != "" {
			output += fmt.Sprintf("    Error: %s\n", entry.Error)
		}

		output += "\n"
	}

	return output, nil
}

File: leave-system/agents/agent_logger.py
"""
Agent Logger for Leave System
Tracks AI agent decision-making process and conversations
"""

import json
import os
from datetime import datetime
from typing import Dict, Any, List, Optional
from pathlib import Path
import logging

logger = logging.getLogger(__name__)


class AgentLogger:
    """
    Logs agent interactions and decision-making process
    Stores logs in JSONL format organized by chat and date
    """

    def __init__(self, log_dir: str = "logs/agent-logs"):
        """
        Initialize agent logger

        Args:
            log_dir: Directory to store agent logs
        """
        self.log_dir = Path(log_dir)
        self.log_dir.mkdir(parents=True, exist_ok=True)

    def log_received_message(
        self,
        chat_id: str,
        sender: str,
        message: str,
        channel: str
    ):
        """Log when a message is received"""
        entry = {
            "timestamp": datetime.now().isoformat(),
            "chat_id": chat_id,
            "stage": "received",
            "sender": sender,
            "channel": channel,
            "message": message,
            "logic": {
                "message_length": len(message),
                "has_content": bool(message.strip())
            }
        }
        self._write_log_entry(chat_id, entry)

    def log_analysis(
        self,
        chat_id: str,
        analysis: Dict[str, Any]
    ):
        """Log message analysis results"""
        entry = {
            "timestamp": datetime.now().isoformat(),
            "chat_id": chat_id,
            "stage": "analysis",
            "logic": analysis
        }
        self._write_log_entry(chat_id, entry)

    def log_decision(
        self,
        chat_id: str,
        decision: Dict[str, Any]
    ):
        """Log agent decision"""
        entry = {
            "timestamp": datetime.now().isoformat(),
            "chat_id": chat_id,
            "stage": "decision",
            "logic": decision
        }
        self._write_log_entry(chat_id, entry)

    def log_response(
        self,
        chat_id: str,
        response: str,
        logic: Dict[str, Any]
    ):
        """Log generated response"""
        entry = {
            "timestamp": datetime.now().isoformat(),
            "chat_id": chat_id,
            "stage": "response",
            "response": response,
            "logic": logic
        }
        self._write_log_entry(chat_id, entry)

    def log_error(
        self,
        chat_id: str,
        stage: str,
        error: str
    ):
        """Log an error"""
        entry = {
            "timestamp": datetime.now().isoformat(),
            "chat_id": chat_id,
            "stage": stage,
            "error": error
        }
        self._write_log_entry(chat_id, entry)

    def log_leave_request_analysis(
        self,
        chat_id: str,
        message: str,
        extracted_info: Dict[str, Any],
        missing_fields: List[str],
        next_action: str
    ):
        """
        Log detailed leave request analysis

        Args:
            chat_id: Chat identifier
            message: Original message text
            extracted_info: Information extracted from the message
            missing_fields: Required fields that are missing
            next_action: What the agent should do next
        """
        # Build detailed analysis text
        analysis_detail = ""

        if extracted_info:
            analysis_detail += "Found information:\n"
            for key, value in extracted_info.items():
                analysis_detail += f"  - {key}: {value}\n"

        if missing_fields:
            analysis_detail += "\nMissing required information:\n"
            for field in missing_fields:
                analysis_detail += f"  - {field}\n"

        analysis_detail += f"\nNext action: {next_action}\n"

        entry = {
            "timestamp": datetime.now().isoformat(),
            "chat_id": chat_id,
            "stage": "leave_request_analysis",
            "message": message,
            "logic": {
                "intent": "leave_request",
                "message_content": message,
                "extracted_info": extracted_info,
                "missing_fields": missing_fields,
                "next_action": next_action,
                "analysis_detail": analysis_detail
            }
        }
        self._write_log_entry(chat_id, entry)

    def _write_log_entry(self, chat_id: str, entry: Dict[str, Any]):
        """Write a log entry to the appropriate file"""
        try:
            # Create daily log file for each chat
            date = datetime.now().strftime("%Y-%m-%d")
            # Sanitize chat_id for filename
            safe_chat_id = chat_id.replace('/', '_').replace('@', '_at_')
            log_file = self.log_dir / f"{safe_chat_id}_{date}.jsonl"

            # Append to log file
            with open(log_file, 'a') as f:
                f.write(json.dumps(entry) + '\n')

        except Exception as e:
            logger.error(f"Failed to write log entry: {e}")

    def get_chat_logs(
        self,
        chat_id: str,
        date: str = None
    ) -> List[Dict[str, Any]]:
        """
        Retrieve logs for a specific chat

        Args:
            chat_id: Chat identifier
            date: Date string (YYYY-MM-DD), defaults to today

        Returns:
            List of log entries
        """
        if date is None:
            date = datetime.now().strftime("%Y-%m-%d")

        safe_chat_id = chat_id.replace('/', '_').replace('@', '_at_')
        log_file = self.log_dir / f"{safe_chat_id}_{date}.jsonl"

        if not log_file.exists():
            return []

        entries = []
        try:
            with open(log_file, 'r') as f:
                for line in f:
                    if line.strip():
                        entries.append(json.loads(line))
        except Exception as e:
            logger.error(f"Failed to read log file: {e}")

        return entries

    def generate_human_readable_log(
        self,
        chat_id: str,
        date: str = None
    ) -> str:
        """
        Generate a human-readable summary of logs

        Args:
            chat_id: Chat identifier
            date: Date string (YYYY-MM-DD), defaults to today

        Returns:
            Human-readable log text
        """
        entries = self.get_chat_logs(chat_id, date)

        if not entries:
            return f"No logs found for {chat_id} on {date or 'today'}\n"

        output = f"=== Agent Logs for {chat_id} ({date or 'today'}) ===\n\n"

        for i, entry in enumerate(entries, 1):
            timestamp = entry.get('timestamp', 'unknown')
            stage = entry.get('stage', 'unknown')

            output += f"[{i}] {timestamp} - Stage: {stage}\n"

            if entry.get('sender'):
                output += f"    Sender: {entry['sender']}\n"

            if entry.get('message'):
                output += f"    Message: {entry['message']}\n"

            if entry.get('logic'):
                output += "    Logic/Analysis:\n"
                logic = entry['logic']
                for key, value in logic.items():
                    if key == 'analysis_detail':
                        output += f"      {value}\n"
                    else:
                        output += f"      {key}: {value}\n"

            if entry.get('response'):
                output += f"    Response: {entry['response']}\n"

            if entry.get('error'):
                output += f"    Error: {entry['error']}\n"

            output += "\n"

        return output
